File: prism/taint.py
"""Conservative intra-function source→sink taint on C text.

CodeQL-shaped lint, not a proof. POINTER functions are in scope.
No flow → skip the function; never emit CLEAN-as-proof.
"""

from __future__ import annotations

import re

from prism import laws
from prism.models import Finding, FunctionInfo

_SOURCES = frozenset({
    "getenv", "fgets", "gets", "scanf", "recv", "read", "fread", "recvfrom",
})
_SINKS = frozenset({
    "system", "popen", "execl", "execv", "strcpy", "sprintf", "memcpy",
    "printf", "fprintf", "strcat", "strncat",
})

_ASSIGN = re.compile(
    r"(?P<lhs>[A-Za-z_]\w*(?:\s*\[[^\]]*\])?)\s*=\s*(?P<rhs>[^;]+)",
)
_SIMPLE_COPY = re.compile(
    r"(?P<dst>[A-Za-z_]\w*)\s*=\s*(?:\([^)]*\)\s*)*(?P<src>[A-Za-z_]\w*)\s*(?:;|$)",
)
_IDENT = re.compile(r"\b([A-Za-z_]\w*)\b")
_STRING = re.compile(r'"([^"\\]|\\.)*"')


def _split_args(argtext: str) -> list[str]:
    args: list[str] = []
    depth = 0
    cur: list[str] = []
    for ch in argtext:
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth = max(0, depth - 1)
        elif ch == "," and depth == 0:
            args.append("".join(cur).strip())
            cur = []
            continue
        cur.append(ch)
    tail = "".join(cur).strip()
    if tail:
        args.append(tail)
    return args


def _sink_arg_indices(name: str) -> tuple[int, ...]:
    if name in {"system", "popen", "execl", "execv", "printf"}:
        return (0,)
    if name in {"sprintf", "fprintf"}:
        return (1,)
    if name in {"strcpy", "memcpy", "strcat", "strncat"}:
        return (0, 1)
    return (0,)


def _arg_is_tainted(arg: str, tainted: set[str]) -> bool:
    stripped = _STRING.sub("", arg)
    ids = {m.group(1) for m in _IDENT.finditer(stripped)}
    return bool(ids & tainted)


def _mark_source_taint(stmt: str, tainted: set[str]) -> None:
    for m in _ASSIGN.finditer(stmt):
        lhs = m.group("lhs").split("[", 1)[0].strip()
        rhs = m.group("rhs")
        if any(re.search(rf"\b{re.escape(src)}\s*\(", rhs) for src in _SOURCES):
            tainted.add(lhs)

    for src in _SOURCES:
        for m in re.finditer(rf"\b{re.escape(src)}\s*\((?P<args>[^)]*)\)", stmt):
            args = _split_args(m.group("args"))
            if not args:
                continue
            if src == "read" and len(args) >= 2:
                tainted.add(args[1].strip().lstrip("&").split("[", 1)[0])
            elif src == "recv" and len(args) >= 2:
                tainted.add(args[1].strip().lstrip("&").split("[", 1)[0])
            elif src == "recvfrom" and len(args) >= 2:
                tainted.add(args[1].strip().lstrip("&").split("[", 1)[0])
            elif src == "fread":
                tainted.add(args[0].strip().lstrip("&").split("[", 1)[0])
            elif src in {"fgets", "gets"}:
                tainted.add(args[0].strip().lstrip("&").split("[", 1)[0])
            elif src == "scanf":
                for arg in args[1:]:
                    tainted.add(arg.strip().lstrip("&*").split("[", 1)[0])


def _mark_copy_taint(stmt: str, tainted: set[str]) -> None:
    m = _SIMPLE_COPY.search(stmt)
    if not m:
        return
    src, dst = m.group("src"), m.group("dst")
    if src in tainted:
        tainted.add(dst)


def _check_sinks(
    stmt: str,
    fn: FunctionInfo,
    line: int,
    tainted: set[str],
) -> Finding | None:
    for m in re.finditer(
        r"\b(?P<name>system|popen|execl|execv|strcpy|sprintf|memcpy|strcat|strncat)\s*\((?P<args>[^)]*)\)",
        stmt,
    ):
        name = m.group("name")
        args = _split_args(m.group("args"))
        for idx in _sink_arg_indices(name):
            if idx >= len(args):
                continue
            if _arg_is_tainted(args[idx], tainted):
                return Finding(
                    stage="taint",
                    status=laws.FAILED,
                    file=fn.file,
                    function=fn.name,
                    line=line,
                    cls="TAINT-SINK",
                    message=f"tainted data reaches {name}()",
                    strength=laws.STRENGTH_FINDS,
                    evidence=stmt.strip(),
                )
    return None


def _iter_statements(body: str, base_line: int) -> list[tuple[int, str]]:
    lines = body.splitlines()
    out: list[tuple[int, str]] = []
    buf: list[str] = []
    start = 0
    for i, raw in enumerate(lines):
        if not buf:
            start = i
        buf.append(raw)
        if ";" in raw or raw.strip().endswith("}"):
            stmt = " ".join(buf).strip()
            if stmt:
                out.append((base_line + start + 1, stmt))
            buf = []
    tail = " ".join(buf).strip()
    if tail:
        out.append((base_line + start + 1, tail))
    return out


def _analyze_function(fn: FunctionInfo) -> Finding | None:
    tainted: set[str] = set()
    for _, name in fn.params:
        if name in {"argv", "envp"}:
            tainted.add(name)

    for line_no, stmt in _iter_statements(fn.body, fn.line):
        _mark_source_taint(stmt, tainted)
        _mark_copy_taint(stmt, tainted)
        hit = _check_sinks(stmt, fn, line_no, tainted)
        if hit:
            return hit
    return None


def run_taint(functions: list[FunctionInfo]) -> list[Finding]:
    out: list[Finding] = []
    for fn in functions:
        hit = _analyze_function(fn)
        if hit:
            out.append(hit)
    return out

File: src/prism/stages/taint.cpp
// Stage taint: source-to-sink taint through assignments and copies (run_taint).
#include "common.hpp"

namespace prism {
namespace fs = std::filesystem;
using namespace stages_detail;

namespace {
std::vector<std::string> taint_split_args(const std::string& argtext) {
    std::vector<std::string> args;
    int depth = 0;
    std::string cur;
    for (char ch : argtext) {
        if (ch == '(') {
            ++depth;
            cur.push_back(ch);
        } else if (ch == ')') {
            depth = std::max(0, depth - 1);
            cur.push_back(ch);
        } else if (ch == ',' && depth == 0) {
            args.push_back(strip(cur));
            cur.clear();
        } else {
            cur.push_back(ch);
        }
    }
    auto tail = strip(cur);
    if (!tail.empty()) args.push_back(tail);
    return args;
}

std::vector<int> sink_arg_indices(const std::string& name) {
    if (name == "system" || name == "popen" || name == "execl" || name == "execv" || name == "printf")
        return {0};
    if (name == "sprintf" || name == "fprintf") return {1};
    if (name == "strcpy" || name == "memcpy" || name == "strcat" || name == "strncat") return {0, 1};
    return {0};
}

bool arg_is_tainted(const std::string& arg, const std::set<std::string>& tainted) {
    static Regex str_re(R"("([^"\\]|\\.)*")");
    std::string stripped;
    std::size_t i = 0;
    for (auto& m : str_re.finditer(arg)) {
        if (m.spans.empty()) continue;
        auto a = static_cast<std::size_t>(std::max(0, m.spans[0].first));
        auto b = static_cast<std::size_t>(std::max(0, m.spans[0].second));
        if (a < i) continue;
        stripped.append(arg, i, a - i);
        i = b;
    }
    stripped.append(arg, i, std::string::npos);
    static Regex ident("\\b([A-Za-z_]\\w*)\\b");
    for (auto& m : ident.finditer(stripped))
        if (tainted.contains(m.group(1))) return true;
    return false;
}

std::vector<std::pair<int, std::string>> iter_statements(const std::string& body, int base_line) {
    std::vector<std::string> lines;
    {
        std::istringstream ss(body);
        std::string ln;
        while (std::getline(ss, ln)) lines.push_back(ln);
    }
    std::vector<std::pair<int, std::string>> out;
    std::vector<std::string> buf;
    int start = 0;
    for (int i = 0; i < static_cast<int>(lines.size()); ++i) {
        if (buf.empty()) start = i;
        buf.push_back(lines[static_cast<std::size_t>(i)]);
        auto raw = lines[static_cast<std::size_t>(i)];
        if (raw.find(';') != std::string::npos || strip(raw).ends_with("}")) {
            auto stmt_s = strip(join_sv(buf, " "));
            if (!stmt_s.empty()) out.emplace_back(base_line + start + 1, stmt_s);
            buf.clear();
        }
    }
    auto tail = strip(join_sv(buf, " "));
    if (!tail.empty()) out.emplace_back(base_line + start + 1, tail);
    return out;
}

void mark_source_taint(const std::string& stmt_s, std::set<std::string>& tainted) {
    static const std::vector<std::string> sources{"getenv", "fgets", "gets", "scanf", "recv", "read", "fread", "recvfrom"};
    static Regex asg("(?P<lhs>[A-Za-z_]\\w*(?:\\s*\\[[^\\]]*\\])?)\\s*=\\s*(?P<rhs>[^;]+)");
    for (auto& m : asg.finditer(stmt_s)) {
        auto lhs = m.named("lhs");
        auto br = lhs.find('[');
        if (br != std::string::npos) lhs = strip(lhs.substr(0, br));
        else lhs = strip(lhs);
        auto rhs = m.named("rhs");
        for (auto& src : sources)
            if (re_search("\\b" + src + "\\s*\\(", rhs)) tainted.insert(lhs);
    }
    for (auto& src : sources) {
        Regex call("\\b" + src + "\\s*\\((?P<args>[^)]*)\\)");
        for (auto& m : call.finditer(stmt_s)) {
            auto args = taint_split_args(m.named("args"));
            if (args.empty()) continue;
            auto take = [](std::string a) {
                a = strip(a);
                while (!a.empty() && (a.front() == '&' || a.front() == '*')) a.erase(a.begin());
                auto br = a.find('[');
                if (br != std::string::npos) a = a.substr(0, br);
                return strip(a);
            };
            if (src == "read" && args.size() >= 2) tainted.insert(take(args[1]));
            else if (src == "recv" && args.size() >= 2) tainted.insert(take(args[1]));
            else if (src == "recvfrom" && args.size() >= 2) tainted.insert(take(args[1]));
            else if (src == "fread") tainted.insert(take(args[0]));
            else if (src == "fgets" || src == "gets") tainted.insert(take(args[0]));
            else if (src == "scanf") {
                for (std::size_t i = 1; i < args.size(); ++i) tainted.insert(take(args[i]));
            }
        }
    }
}

void mark_copy_taint(const std::string& stmt_s, std::set<std::string>& tainted) {
    static Regex re("(?P<dst>[A-Za-z_]\\w*)\\s*=\\s*(?:\\([^)]*\\)\\s*)*(?P<src>[A-Za-z_]\\w*)\\s*(?:;|$)");
    auto m = re.search_match(stmt_s);
    if (!m) return;
    if (tainted.contains(m->named("src"))) tainted.insert(m->named("dst"));
}

std::optional<Finding> analyze_taint(const FunctionInfo& fn) {
    std::set<std::string> tainted;
    for (auto& [t, name] : fn.params)
        if (name == "argv" || name == "envp") tainted.insert(name);
    for (auto& [line_no, stmt_s] : iter_statements(fn.body, fn.line)) {
        mark_source_taint(stmt_s, tainted);
        mark_copy_taint(stmt_s, tainted);
        static Regex sink(
            "\\b(?P<name>system|popen|execl|execv|strcpy|sprintf|memcpy|strcat|strncat)\\s*\\((?P<args>[^)]*)\\)");
        for (auto& m : sink.finditer(stmt_s)) {
            auto name = m.named("name");
            auto args = taint_split_args(m.named("args"));
            for (int idx : sink_arg_indices(name)) {
                if (idx >= static_cast<int>(args.size())) continue;
                if (arg_is_tainted(args[static_cast<std::size_t>(idx)], tainted)) {
                    auto f = make_find("taint", laws::FAILED, fn, "TAINT-SINK",
                                       "tainted data reaches " + name + "()", laws::STRENGTH_FINDS);
                    f.line = line_no;
                    f.evidence = strip(stmt_s);
                    return f;
                }
            }
        }
    }
    return std::nullopt;
}

}  // namespace

std::vector<Finding> run_taint(const std::vector<FunctionInfo>& functions) {
    std::vector<Finding> out;
    for (auto& fn : functions)
        if (auto hit = analyze_taint(fn)) out.push_back(*hit);
    return out;
}

}  // namespace prism
